File: dmoj/vpex1p4.cpp
#include <iostream>
#include <bits/stdc++.h>
using namespace std;
#pragma GCC optimize("Ofast")
#pragma GCC target("avx2")
#define ll unsigned long long


class edge{
    public:
    ll u;
    ll v;
    ll weight;

    edge(ll u, ll v, ll weight) : u(u), v(v), weight(weight) {}
    bool operator<(const edge& other) const {
        return weight < other.weight;
    }
};

vector<edge> adj[100005];
vector<ll> euler;
ll depth[100005];
ll dist[100005];
ll in[100005];
ll day[2][1000000];
pair<ll, ll> sparse[2*100000+5][21];
    
void dfs(edge cur){
    euler.push_back(cur.v);
    if(in[cur.v] == -1) in[cur.v] = euler.size()-1;

    for(edge next : adj[cur.v]){
        if(euler[euler.size()-1] != cur.v) euler.push_back(cur.v);
        if(depth[cur.v] + 1 < depth[next.v]){
            depth[next.v] = depth[cur.v] + 1;
            if(in[next.v] == -1) in[next.v] = euler.size();
            dfs(next);
            euler.push_back(next.v);
        }
    }
}

ll query(ll L, ll R){
    if(L > R) swap(L, R);
    ll j = (ll)log2(R - L + 1);
 
    if (sparse[L][j].first <= sparse[R - (1 << j) + 1][j].first) return sparse[L][j].second;
    else return sparse[R - (1 << j) + 1][j].second;
}

int main() {
    cin.tie(NULL);
    ios_base::sync_with_stdio(false);
    
    ll N, D;
    cin >> N >> D;

    for(ll i = 0; i < N-1; ++i){
        ll a, b, c;
        cin >> a >> b >> c;
        --a; --b;
        adj[a].push_back(edge(a, b, c));
        adj[b].push_back(edge(b, a, c));
    }


    fill(depth, depth+N, LONG_LONG_MAX);
    fill(in, in+N, -1);
    depth[0] = 0;
    dfs(edge(0, 0, 0));

    for(ll i = 0; i < (ll)euler.size(); ++i) 
        sparse[i][0] = make_pair(depth[euler[i]], euler[i]);
    for(ll j = 1; j <= (ll)log2(euler.size()); ++j) {
        for(ll i = 1; i + (1 << j) - 1 <= (ll)euler.size(); i++) {
            if (sparse[i][j - 1].first < sparse[i + (1 << (j - 1))][j - 1].first) sparse[i][j] = sparse[i][j - 1];
            else sparse[i][j] = sparse[i + (1 << (j - 1))][j - 1];
        }
    }


    fill(dist, dist+N, LONG_LONG_MAX);
    priority_queue<edge> pq;
    pq.push(edge(0, 0, 0));
    dist[0] = 0;
    while(!pq.empty()){
        edge cur = pq.top();
        pq.pop();
        
        for(edge next : adj[cur.v]){
            if(dist[cur.v] + next.weight < dist[next.v]){
                dist[next.v] = dist[cur.v] + next.weight;
                pq.push(next);
            }
        }
    }


    for(ll i = 0; i < D; ++i){
        ll x, y;
        cin >> x >> y;
        --x; --y;

        day[0][i] = x;
        day[1][i] = y;
    }

    ll dp[D][2];
    dp[0][0] = dist[day[1][0]] + dist[day[0][0]]+dist[day[1][0]]-2*dist[query(in[day[0][0]], in[day[1][0]])];
    dp[0][1] = dist[day[0][0]] + dist[day[0][0]]+dist[day[1][0]]-2*dist[query(in[day[0][0]], in[day[1][0]])]; 

    for(int i = 1; i < D; ++i){
        dp[i][0] = min(
            dp[i-1][0] + dist[day[0][i-1]]+dist[day[1][i]]-2*dist[query(in[day[0][i-1]], in[day[1][i]])], 
            dp[i-1][1] + dist[day[1][i-1]]+dist[day[1][i]]-2*dist[query(in[day[1][i-1]], in[day[1][i]])]
        ) + dist[day[0][i]] + dist[day[1][i]] - 2*dist[query(in[day[0][i]], in[day[1][i]])];

        dp[i][1] = min(
            dp[i-1][0] + dist[day[0][i-1]]+dist[day[0][i]]-2*dist[query(in[day[0][i-1]], in[day[0][i]])], 
            dp[i-1][1] + dist[day[1][i-1]]+dist[day[0][i]]-2*dist[query(in[day[1][i-1]], in[day[0][i]])]
        ) + dist[day[0][i]] + dist[day[1][i]] - 2*dist[query(in[day[0][i]], in[day[1][i]])];
    }

    std::cout << min(dp[D-1][0], dp[D-1][1]) << "\n";
}

File: dmoj/vpex1p4.java
import java.util.*;
import java.io.*;
	
public class vpex1p4 {
	
	public static class pair<K, V> {
		K key;
		V value;
		
		public pair(K key, V value) {
			this.key = key;
			this.value = value;
		}
		
		public K first() {
			return key;
		}
		
		public V second() {
			return value;
		}
	}
	
	public static class edge implements Comparable<edge>{
		int u;
	    int v;
	    int weight;

	    public edge(int u, int v, int weight) {
	    	this.u = u;
	    	this.v = v;
	    	this.weight = weight;
	    }
	    
	    public int compareTo(edge other){
	        return this.weight - other.weight;
	    }
	}

	static ArrayList<edge>[] adj = new ArrayList[100005];
	static ArrayList<Integer> euler = new ArrayList<>();
	static int[] depth = new int[100005];
	static long[] dist = new long[100005];
	static int[] in = new int[100005];
	static pair<Integer, Integer>[][] sparse = new pair[2*100000+5][18];
	    
	public static void dfs(edge cur){
	    euler.add(cur.v);

	    for(edge next : adj[cur.v]){
	        if(depth[cur.v] + 1 < depth[next.v]){
	            depth[next.v] = depth[cur.v] + 1;
	            if(in[next.v] == -1) in[next.v] = euler.size();
	            dfs(next);
	            euler.add(next.v);
	        }
	    }
	}

	public static int query(int L, int R){
	    int j = (int)log2(R - L + 1);
	 
	    if (sparse[L][j].first() <= sparse[R - (1 << j) + 1][j].first()) return sparse[L][j].second();
	    else return sparse[R - (1 << j) + 1][j].second();
	}

	public static int log2(int n) {
		return (int)(Math.log(n)/Math.log(2));
	}

	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;

        st = new StringTokenizer(br.readLine());
        int N = Integer.parseInt(st.nextToken()), D = Integer.parseInt(st.nextToken());

        for(int i = 0; i < N; ++i) {
        	adj[i] = new ArrayList<>();
        }
	    for(int i = 0; i < N-1; ++i){
	    	st = new StringTokenizer(br.readLine());
	        int a = Integer.parseInt(st.nextToken()), b = Integer.parseInt(st.nextToken()), c = Integer.parseInt(st.nextToken());
	        --a; --b;
	        adj[a].add(new edge(a, b, c));
	        adj[b].add(new edge(b, a, c));
	    }


	    Arrays.fill(depth, Integer.MAX_VALUE);
	    depth[0] = 0;
	    dfs(new edge(0, 0, 0));

	    for(int i = 0; i < sparse.length; ++i) {
	    	for(int j = 0; j < sparse[0].length; ++j) {
	    		sparse[i][j] = new pair(0, 0);
	    	}
	    }
	    for(int i = 1; i < euler.size(); ++i) 
	        sparse[i][0] = new pair(depth[euler.get(i)], euler.get(i));
	    for(int j = 1; j <= log2(euler.size()); ++j) {
	        for(int i = 1; i + (1 << j) - 1 <= euler.size(); i++) {
	            if (sparse[i][j - 1].first() < sparse[i + (1 << (j - 1))][j - 1].first()) sparse[i][j] = sparse[i][j - 1];
	            else sparse[i][j] = sparse[i + (1 << (j - 1))][j - 1];
	        }
	    }


	    Arrays.fill(dist, Long.MAX_VALUE);
	    PriorityQueue<edge> pq = new PriorityQueue<>();
	    pq.add(new edge(0, 0, 0));
	    dist[0] = 0;
	    while(!pq.isEmpty()){
	        edge cur = pq.poll();
	        
	        for(edge next : adj[cur.v]){
	            if(dist[cur.v] + next.weight < dist[next.v]){
	                dist[next.v] = dist[cur.v] + next.weight;
	                pq.add(next);
	            }
	        }
	    }


	    int cur = 0;
	    long ans = 0;
	    for(int i = 0; i < D; ++i){
	    	st = new StringTokenizer(br.readLine());
	        int x = Integer.parseInt(st.nextToken()), y = Integer.parseInt(st.nextToken());
	        --x; --y;

	        long distx, disty, distxy;
	        if(in[x] < in[cur]){
	            distx = dist[x] + dist[cur] - 2*dist[query(x, cur)];
	        }else{
	            distx = dist[x] + dist[cur] - 2*dist[query(cur, x)];
	        }
	        
	        if(in[y] < in[cur]){
	            disty = dist[y] + dist[cur] - 2*dist[query(y, cur)];
	        }else{
	            disty = dist[y] + dist[cur] - 2*dist[query(cur, y)];
	        }

	        if(in[x] < in[y]){
	            distxy = dist[x] + dist[y] - 2*dist[query(x, y)];
	        }else{
	            distxy = dist[x] + dist[y] - 2*dist[query(y, x)];
	        }

	        ans += Math.min(distx + distxy, disty + distxy);
	        if(distx + distxy < disty + distxy) cur = y;
	        else cur = x;
	    }

	    System.out.println(ans);
	}
}
